Allow wrapper function generation for functional macros

Abestanis opened the pull request:
This allows to register functional macros for which a wrapper function will be generated. Currently it is not possible to call functional macros at all. This makes it possible.

Macros don't have a defined argument or return types, which is why they have to be specified explicitly. Some macros can be called with different argument types. This patch only supports creating a wrapper function for one specific set of argument and return types. Wrappers are only generated for configured macros, all others are ignored.

### Example
With this change we can expose the following C macros:

```c
#define MACRO_WITH_NO_ARGUMENTS printf("Test")
#define MACRO_WITH_NO_ARGUMENTS_AND_RETURN_VALUE() 0
#define MACRO_WITH_ONE_ARGUMENT_AND_RETURN_VALUE(value) ((float) (value))
#define MACRO_WITH_TWO_ARGUMENTS_AND_RETURN_VALUE(condition, value) ((condition) ? ((float) (value)) : 0)
```

To generate wrappers for these macros, we have to eiter configure them via command line arguments:
```text
--macro-function "MACRO_WITH_NO_ARGUMENTS"
--macro-function "u8 MACRO_WITH_NO_ARGUMENTS_AND_RETURN_VALUE"
--macro-function "f32 MACRO_WITH_ONE_ARGUMENT_AND_RETURN_VALUE(u32)"
--macro-function "f32 MACRO_WITH_TWO_ARGUMENTS_AND_RETURN_VALUE(bool, u32)"
```
… or via the `Builder`:
```rust
Builder::default()
     .macro_function("MACRO_WITH_NO_ARGUMENTS", FunctionType::new::<(), ()>())
     .macro_function("MACRO_WITH_NO_ARGUMENTS_AND_RETURN_VALUE", FunctionType::new::<u8, ()>())
     .macro_function("MACRO_WITH_ONE_ARGUMENT_AND_RETURN_VALUE", FunctionType::new::<f32, u32>())
     .macro_function("MACRO_WITH_TWO_ARGUMENTS_AND_RETURN_VALUE", FunctionType::new::<f32, (bool, u32)>())
```

On the Rust side we can then call the macros as normal functions:
```rust
unsafe {
    ffi::MACRO_WITH_NO_ARGUMENTS();
    let value = ffi::MACRO_WITH_NO_ARGUMENTS_AND_RETURN_VALUE();
    let value = ffi::MACRO_WITH_ONE_ARGUMENT_AND_RETURN_VALUE(42);
    let value = ffi::MACRO_WITH_TWO_ARGUMENTS_AND_RETURN_VALUE(true, 42);
}
```

Abestanis committed:
Generate wrapping functions for functional macros

Abestanis committed:
Add tests for macro wrap function generation

Abestanis committed:
Add changes to Changelog

Abestanis commented:
r? @emilio

bors-servo commented:
:umbrella: The latest upstream changes (presumably f44a66578dfcc74da290b9707e0cbb290b5ceeb4) made this pull request unmergeable. Please resolve the merge conflicts.

Abestanis committed:
Make sure macro wrapper functions have the correct link name

bors-servo commented:
:umbrella: The latest upstream changes (presumably 251dec94c8c7c6413bdaec0c40ff2acecc42e660) made this pull request unmergeable. Please resolve the merge conflicts.

pvdrz commented:
I like this feature but I think it needs more bikeshedding.

One issue I see is that having to write the name of the macro itself and the type makes it dangerously close to just creating a wrapper by hand.

If I had a macro in `input.h`:
```c
#define INC(value) (value + 1)
```
I could very well create a `wrapper.h` with these contents:
```c
#include "input.h"

int inc(int val) {
    return INC(val)
}
```
And pass it to bindgen. You could argue that just having to do `--macro-function="int INC(int)"` is shorter but it doesn't feel that much shorter to be honest and the amount of complexity adding to bindgen to do it doesn't feel like a good trade-off to me.

Abestanis commented:
Writing the wrapper manually in a header and passing it to bindgen does create a binding, but the function still needs to be included in the final compilation unit, so we also need a source file that includes the header and add it to the build. Otherwise the function referenced by bindgen is not actually present in the library.

I'm also not a fan of having to specify every macro function type, but I'm not aware of any solution on how to avoid doing that and at least this way I can avoid writing C code, which is why this feature is worth it in my opinion.

bors-servo commented:
:umbrella: The latest upstream changes (presumably 1d2b579d1c3180e93089ffb1379ee11fac0cccc0) made this pull request unmergeable. Please resolve the merge conflicts.

pvdrz commented:
> Writing the wrapper manually in a header and passing it to bindgen does create a binding, but the function still needs to be included in the final compilation unit, so we also need a source file that includes the header and add it to the build. Otherwise the function referenced by bindgen is not actually present in the library.

That's a good point, so basically by piggybacking on the `--wrap-static-fns` video we get the source C file so the user only has to compile it instead of writing it by hand. That's pretty neat! 

> I'm also not a fan of having to specify every macro function type, but I'm not aware of any solution on how to avoid doing that and at least this way I can avoid writing C code, which is why this feature is worth it in my opinion.

Yeah I don't have a better solution for this either. I'll give you a proper review in a moment

pvdrz commented on bindgen-cli/options.rs:
```diff
@@ -362,14 +363,17 @@ struct BindgenCommand {
+    /// Create a wrapper function for the macro. The MACRO value must be of the shape
+    /// `[<return type>] <macro name>[(<comma separated list of arguments>)]`.
+    #[arg(long, value_name = "MACRO")]
+    macro_function: Option<Vec<String>>,
```
I suspect this should require the `experimental` feature as it leverages the infrastructure for `--wrap-static-fns` which is itself experimental

pvdrz commented on bindgen-cli/options.rs:
```diff
@@ -504,8 +508,9 @@ where
         with_derive_custom_union,
         wrap_static_fns,
-        wrap_static_fns_path,
+        native_code_generation_path,
```
I don't love the `native` here, maybe just saying `c_code_generation_path` or `wrapper_code_generation_path` is better?

pvdrz commented on bindgen-cli/options.rs:
```diff
@@ -1084,3 +1136,28 @@ where
 }
+
+/// Parse a [`CTypeKind`] from the given command line `input`.
+fn parse_c_type(input: &str) -> Result<CTypeKind, String> {
```
is this really necessary? It might sound lazy but to be honest it might be easier to just let the user type whatever they want as a type and propagate those strings until they are written to the wrapper code file. This also gives you the "advantage" of supporting any type the user wants as long as it is defined in the headers.

Abestanis replied:
I am actually a bit stuck on this. I am trying to support calling these macros with C types, which can't be referenced from build scripts:
```c
typedef struct {
    int inner;
} my_type;

#define ACCESS_INNER(value) value.inner
```

I want to be able to specify that the `ACCESS_INNER` macro takes a `my_type`. To do that I was trying to switch everything over to just accept strings, but in `Function::parse` I need to generate a `Function` object which eventually needs a `TypeKind` for the return type and the argument types.

So my question is, is there a way to construct a `Function` that doesn't require me to parse the types, so I can just pass the strings through?

pvdrz commented on bindgen/codegen/mod.rs:
```diff
@@ -4236,6 +4251,9 @@ impl CodeGenerator for Function {
             let name = canonical_name.clone() + ctx.wrap_static_fns_suffix();
             attributes.push(attributes::link_name::<true>(&name));
+        } else if is_function_macro && !has_link_name_attr {
+            let name = canonical_name.clone() + "__macro";
```
I think this `__macro` suffix should be configurable, maybe reusing the suffix for static wrappers is a good idea, you could rename it to `wrapper_suffix` or something.

pvdrz commented on bindgen/codegen/mod.rs:
```diff
@@ -4787,6 +4810,7 @@ pub(crate) mod utils {
 
         let mut code = Vec::new();
 
+        writeln!(code, "#include <stdint.h>\n#include <stdbool.h>")?;
```
I wouldn't add this here as it would break people's code that don't have these libraries or have their system path messed up (like in some windows casses I've seen). This goes in the same direction as treating the argument and return types provided by the user be just strings, if they use `bool` they have to include it in the headers themselves.

Abestanis replied:
Hm, I agree that unconditionally adding those is not a good idea, but the problem is that people can't include the headers themselves, since the wrapper C file only includes the header file where the macro is defined over which the user usually has no control and which might not have the the required headers included.

Should we add an additional option allow passing in include files for the generated C code? Or do you have a better idea?

pvdrz commented on bindgen-tests/tests/headers/function_macros.h:
```diff
@@ -0,0 +1,15 @@
+// bindgen-flags: --macro-function "SIMPLE" --macro-function "INDIRECT_SIMPLE" --macro-function "f32 COMPLEX(u32)" --macro-function "f32 INDIRECT_COMPLEX(u32)" --macro-function "f32 CONDITIONAL_COMPLEX(bool, u32)"
```
I have mixed feelings about the syntax for the `--macro-function` argument. Usually when we have these options where you need to match an item by name and need extra info for it we use the `--option=<ITEM_NAME>=<EXTRA_STUFF>` syntax like in the `--override-abi` case.

Maybe having `--macro-function=<MACRO_NAME>=<WRAPPER_TYPE>` is a bit more in style with the rest of bindgen.

This also opens the gates to `<MACRO_NAME>` being a regex in the future.

Abestanis replied:
Hm, the benefit of the the current representation is that is matches the macro definition with the rust types for argument names. For me that looks very readable, but we can try to match the existing command line style.

Did you have an idea for the `<EXTRA_STUFF>` syntax? Should we try to match the syntax used with the builder pattern like this?
`--macro-function=SIMPLE --macro-function=COMPLEX="f32 (u32)" --macro-function=INDIRECT_COMPLEX="f32 (u32)" --macro-function=CONDITIONAL_COMPLEX="f32 (bool, u32)"`

pvdrz reviewed (changes requested):
I'm done with the review. Other than the comments I wrote in the changes I think we might need a different name for this feature as `ParseCallbacks::func_macro` sounds almost like the same, maybe a more explicit `--wrap-func-macro` is better? IDK. The other thing I'd ask you is avoid editing the style of the `CHANGELOG.md` so you don't get more merge conflicts.

reitermarkus commented:
@Abestanis, does https://github.com/rust-lang/rust-bindgen/pull/2369 not solve your use case?

Abestanis commented:
@reitermarkus #2369 is very interesting and incredibly cool. Thank you for putting so much effort into it.  I can indeed use it for my specific usecase of the one specific macro I am trying to make accessible, but I think there is an issue with the approach of trying to replicate the C macro in Rust:

C and macros have a low of quirks and Rust is different to C and macros in many ways. There will always be some macro which can't yet be represented by #2369 and needs some new feature. Specifically when it comes to types and operations between types, C is a lot more lenient (because of implicit type conversions) while Rust basically disallows most operations between types. This means that many macros loose their ability to be called with different types.

I think the following (terrible, but) valid macro is going to be difficult to represent with #2369, even though I'm sure there are workarounds for this:
```c
#define WEIRD_IS_ONE(x) (x == (uint16_t) 1 || x == (char) '1')
```
The wrapping approach supports all macro types out of the box, as long as the arguments and return types can be sent across ffi.

In my specific use case I am trying to make the following macro available:
  ```c
  typedef long BaseType_t;
  #define pdFALSE ( ( BaseType_t ) 0 )
  #define portEND_SWITCHING_ISR( xSwitchRequired ) do { if( xSwitchRequired != pdFALSE ) portYIELD(); } while( 0 )
  #define portYIELD_FROM_ISR( x ) portEND_SWITCHING_ISR( x )
  ```
  With the wrapping approach, I can define a wrapper like this
  ```rust
  bindgen::Builder::default()
      .macro_function("portYIELD_FROM_ISR", FunctionType::new::<(), bool>())
      .generate();
  ```
  and call it like this:
  ```
  pub fn yield_from_isr(should_do_isr_context_switch: bool) {
      unsafe { ffi::portYIELD_FROM_ISR(should_do_isr_context_switch) }
  }
  ```
  Even though ` xSwitchRequired` and `pdFALSE` have different types, C allows the comparison to work as expected.

  With #2369, the following binding macro is generated:
  ```rust
  pub use __cmacro__portEND_SWITCHING_ISR as portEND_SWITCHING_ISR;
  #[doc(hidden)]
  #[macro_export]
  macro_rules! __cmacro__portYIELD_FROM_ISR {
      ($ x : expr) => {
          if $x != 0u8 as BaseType_t {
              vPortYield();
          }
      };
  }
  ```
  I can not call this macro with a `bool` argument type any more, because a `<bool> != <cty::c_long>` is not allowed in Rust.
  Instead I have to work around this issue by explicitly casting the only allowed argument type:
  ```rust
  pub fn yield_from_isr(should_do_isr_context_switch: bool) {
      unsafe { ffi::portYIELD_FROM_ISR!(BaseType_t::from(should_do_isr_context_switch)) }
  }
  ```

So I still think the wrapping approach has some value and could exist together with #2369, but we have to make sure that they don't interfere with each other.

Abestanis committed:
Make the macro-function argument experimental

Abestanis committed:
Rename native_code_generation_path to wrapper_code_generation_path

Abestanis committed:
Don't use let...else statement

Abestanis committed:
Rename wrap_static_fns_suffix and use it for functional macros

Abestanis commented:
@pvdrz Sorry for the long period of silence, I have addressed your feedback.

> we might need a different name for this feature as ParseCallbacks::func_macro sounds almost like the same, maybe a more explicit --wrap-func-macro is better?

I'm open for better names, although the `ParseCallbacks::func_macro` is referring to the exact same concept of a functional macro, so I don't think the two are in conflict. Both allow handling functional macros in some way.

> The other thing I'd ask you is avoid editing the style of the CHANGELOG.md so you don't get more merge conflicts.

Yeah, sorry about that, that was not a good idea. I removed all the formatting changes there except the trailing whitespaces, because my editor really wants to remove them when saving. I hope it's ok to leave these changes in there, but If needed I can try to remove them from the commit.

reitermarkus commented:
> In my specific use case I am trying to make the following macro available:
> ```c
> typedef long BaseType_t;
> #define pdFALSE ( ( BaseType_t ) 0 )
> #define portEND_SWITCHING_ISR( xSwitchRequired ) do { if( xSwitchRequired != pdFALSE ) > portYIELD(); } while( 0 )
> #define portYIELD_FROM_ISR( x ) portEND_SWITCHING_ISR( x )
> ```

FreeRTOS is exactly what started my PR, actually. You may want to take a look at my fork of `freertos-rust` here, which is using the `cmacro` branch of my `bindgen` fork to generate (almost) all FreeRTOS bindings: https://github.com/reitermarkus/freertos-rust

But yes, the `portEND_SWITCHING_ISR` macro cannot be easily generated automatically since it accesses a global variable which is defined in a `.c` file. It's definitely possible to add that functionality similar to how types are looked up, but for global variables.

reitermarkus commented:
@Abestanis, I just found a bug that prevented the `portEND_SWITCHING_ISR` from being parsed correctly and hence from being generated, it should be possible now with my PR.

Abestanis commented:
@reitermarkus I would like to give your MR another try, but it seems like it is no longer compiling.

bors-servo commented:
:umbrella: The latest upstream changes (presumably 285eb56cf5a02ee778880b8e1b7bb2e8118a5a88) made this pull request unmergeable. Please resolve the merge conflicts.

Abestanis commented:
@reitermarkus Thank you! ❤️

CGMossa commented:
This looks great!
